sdk/identity/azure-identity-v2/src/main/java/com/azure/identity/v2/AuthorizationCodeCredential.java

```diff
@@ -0,0 +1,130 @@
+import com.azure.identity.v2.implementation.models.MsalAuthenticationAccount;
+import com.azure.identity.v2.implementation.models.MsalToken;
+import com.azure.identity.v2.implementation.models.PublicClientOptions;
+import com.azure.identity.v2.implementation.util.LoggingUtil;
+import com.azure.v2.core.credentials.TokenCredential;
+import com.azure.v2.core.credentials.TokenRequestContext;
+import io.clientcore.core.credentials.oauth.AccessToken;
+import io.clientcore.core.instrumentation.logging.ClientLogger;
+import io.clientcore.core.utils.CoreUtils;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * <p>Authorization Code authentication in Azure is a type of authentication mechanism that allows users to
+ * authenticate with <a href="https://learn.microsoft.com/entra/fundamentals/">Microsoft Entra ID</a>
+ * and obtain an authorization code that can be used to request an access token to access
+ * Azure resources. It is a widely used authentication mechanism and is supported by a wide range of Azure services
+ * and applications. It provides a secure and scalable way to authenticate users and grant them access to Azure
+ * resources.
+ * The AuthorizationCodeCredential authenticates a user or an application and acquires a token with the configured
+ * authorization code and the redirectURL where authorization code was received.</p>
+ *
+ * <p><strong>Sample: Construct AuthorizationCodeCredential</strong></p>
+ *
+ * <p>The following code sample demonstrates the creation of a {@link AuthorizationCodeCredential},
+ * using the {@link AuthorizationCodeCredentialBuilder} to configure it.
+ * The {@code authorizationCode}, {@code redirectUrl} and {@code clientId} are required to be configured to create
+ * {@link AuthorizationCodeCredential}. Once this credential is created, it may be passed into the builder of many of
+ * the Azure SDK for Java client builders as the 'credential' parameter.</p>
+ *
+ * <!-- src_embed com.azure.identity.credential.authorizationcodecredential.construct -->
+ * <pre>
+ * TokenCredential authorizationCodeCredential = new AuthorizationCodeCredentialBuilder&#40;&#41;.authorizationCode&#40;
+ *         &quot;&#123;authorization-code-received-at-redirectURL&#125;&quot;&#41;
+ *     .redirectUrl&#40;&quot;&#123;redirectUrl-where-authorization-code-is-received&#125;&quot;&#41;
+ *     .clientId&#40;&quot;&#123;clientId-of-application-being-authenticated&quot;&#41;
+ *     .build&#40;&#41;;
+ * </pre>
+ * <!-- end com.azure.identity.credential.authorizationcodecredential.construct -->
+ *
+ * @see com.azure.identity
+ * @see AuthorizationCodeCredentialBuilder
+ */
+public class AuthorizationCodeCredential implements TokenCredential {
+    private static final ClientLogger LOGGER = new ClientLogger(AuthorizationCodeCredential.class);
+
+    private final ConfidentialClient confidentialClient;
+    private final PublicClient publicClient;
```

Comment:
This is an odd name even though it's in implementation package.

Reply:
It maps to PublicClient flow of msal, Possibly we can prefix msal to it.

Reply:
in this space Public and Confidential are common enough terms I'd prefer to leave it. Let's just make sure the javadoc is explanatory (link to the flows?)